walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.LinkedList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+  private List<TimeRange> busyTimes = new ArrayList<>();
+  private List<TimeRange> freeTimes = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    busyTimes = getBusyTimes(events, request);
+    if (busyTimes.size() > 0) {
+      return (getFreeTimes(findOverlappingTimes(), request));
+    }
+    return (getFreeTimes(busyTimes, request));
+  }
+
+  // Compare attendees of given events with attendees for requested meeting to determine conflicts
+  private List<TimeRange> getBusyTimes(Collection<Event> events, MeetingRequest request) {
+
+    Collection<String> attendees = request.getAttendees();
+    for (Event event : events) {
+      if(!(Collections.disjoint(event.getAttendees(), attendees))) {
+        busyTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    return busyTimes;
+  }
+
+  // Returns list of consolidated busy time ranges accounting for overlapping events
+  private List<TimeRange> findOverlappingTimes() {
+    List<TimeRange> overlappingTimes = new ArrayList<>();
+    // Variables used to keep track of start and end times of previous relevant busy time blocks
+    int prevStart = busyTimes.get(0).start();
+    int prevEnd = busyTimes.get(0).end();
+    for (int i = 1; i < busyTimes.size(); i++) {
+      int currentStart = busyTimes.get(i).start();
+      int currentEnd = busyTimes.get(i).end();
+      if (currentStart > prevEnd) {
+        overlappingTimes.add(TimeRange.fromStartEnd(prevStart, prevEnd, false));
+        prevStart = currentStart;
+        prevEnd = currentEnd;
+      }
+      else if (currentEnd > prevEnd) {
+        prevEnd = currentEnd;
+      }
+    }
+    overlappingTimes.add(TimeRange.fromStartEnd(prevStart, prevEnd, false));
+    return overlappingTimes;
+  }
+  
+  // Gets list of time slots where no attendess have any meetings >= duration time
+  private Collection<TimeRange> getFreeTimes(List<TimeRange> overlappingTimes, MeetingRequest request) {
+    // Variables used to keep track of start and end times of previous relevant busy time blocks
+    int prevStart = 0;
+    int prevEnd = 0;
+    for (TimeRange timeRange : overlappingTimes) {
+      int currentStart = timeRange.start();
+      int currentEnd = timeRange.end();
+      if (currentStart == 0) {
```

Comment:
Why 0?

Reply:
This is the case where the first event starts at 12:00 AM, so we don't need to account for a free block prior to the first event. I added a comment to the code. Let me know if it needs more clarification.